unity/Assets/Scripts/ColorChanger.cs

```diff
@@ -1,84 +1,287 @@
+            ["Bread"] = breadMaterials,
+            ["Pan"] = panMaterials,
+            ["PanDecal"] = panDecalMaterials,
+            ["CoffeeMachine"] = coffeeMachineMaterials,
+        };
+
+        // makes indexing into them faster
+        materialGroups = new Dictionary<string, HashSet<Material>> {
+            ["train"] = new HashSet<Material>(rawTrainMaterials),
+            ["val"] = new HashSet<Material>(rawValMaterials),
+            ["test"] = new HashSet<Material>(rawTestMaterials),
+            ["robothor"] = new HashSet<Material>(rawRobothorMaterials),
+            ["kitchen"] = new HashSet<Material>(rawKitchenMaterials),
+            ["livingroom"] = new HashSet<Material>(rawLivingRoomMaterials),
+            ["bedroom"] = new HashSet<Material>(rawBedroomMaterials),
+            ["bathroom"] = new HashSet<Material>(rawBathroomMaterials)
+        };
 
-        allMaterials = targetMaterials.Concat(backgroundMaterials).Concat(furnitureMaterials).Concat(quickMaterials).ToArray();
+        // cache all the original values
+        origColors = new Dictionary<string, Color[]>();
+        origTextures = new Dictionary<string, Texture[]>();
+        foreach (KeyValuePair<string, Material[]> materialGroup in materials) {
+            Color[] groupColors = new Color[materialGroup.Value.Length];
+            Texture[] groupTextures = new Texture[materialGroup.Value.Length];
+            for (int i = 0; i < materialGroup.Value.Length; i++) {
+                Material mat = materialGroup.Value[i];
+                if (mat.HasProperty("_Color")) {
+                    groupColors[i] = mat.color;
+                }
+                if (mat.HasProperty("_MainTex")) {
+                    groupTextures[i] = mat.mainTexture;
+                }
+            }
+            origColors[materialGroup.Key] = groupColors;
+            origTextures[materialGroup.Key] = groupTextures;
+        }
+    }
+
+    private void swapMaterial(Material mat1, Material mat2) {
+        if (mat1.HasProperty("_MainTex") && mat2.HasProperty("_MainTex")) {
+            Texture tempTexture = mat1.mainTexture;
+            mat1.mainTexture = mat2.mainTexture;
+            mat2.mainTexture = tempTexture;
+        }
+
+        if (mat1.HasProperty("_Color") && mat2.HasProperty("_Color")) {
+            Color tempColor = mat1.color;
+            mat1.color = mat2.color;
+            mat2.color = tempColor;
+        }
+    }
+
+    private void shuffleMaterials(Material[] materialGroup) {
+        for (int n = materialGroup.Length - 1; n >= 0; n--) {
+            int i = Random.Range(0, n + 1);
+            swapMaterial(materialGroup[i], materialGroup[n]);
+        }
+    }
+
+    private void shuffleMaterials(List<Material> materialGroup) {
+        for (int n = materialGroup.Count - 1; n >= 0; n--) {
+            int i = Random.Range(0, n + 1);
+            swapMaterial(materialGroup[i], materialGroup[n]);
+        }
+    }
+
+    public int RandomizeMaterials(
```

Comment:
Can you add the option to pass in a randomSeed so that configurations can be repeated? For all the other random actions we offer the ability to pass in a seed for this reason.

Reply:
Yes, I don't think you actually need the `shuffleMaterials` function. At some point in the past I added the `ListExtensions.cs` file which adds a couple of different `.Shuffle_` methods (in-place shuffling) with various forms of seeding to the `List` class.

Reply:
@Lucaweihs, I don't think this shuffling method would work, because the swaps aren't just changing the order in the List/array, they're instead changing the underlying properties with:
```c#
    private void swapMaterial(Material mat1, Material mat2) {
        if (mat1.HasProperty("_MainTex") && mat2.HasProperty("_MainTex")) {
            Texture tempTexture = mat1.mainTexture;
            mat1.mainTexture = mat2.mainTexture;
            mat2.mainTexture = tempTexture;
        }

        if (mat1.HasProperty("_Color") && mat2.HasProperty("_Color")) {
            Color tempColor = mat1.color;
            mat1.color = mat2.color;
            mat2.color = tempColor;
        }
    }
```

Reply:
> Can you add the option to pass in a randomSeed so that configurations can be repeated?

I suppose. Although we won't be able to maintain the random seed in a backwards compatible way, since new materials will presumably be continuously added. So, this backwards seed can only be guaranteed to be reproducible when using the same build.

Reply:
I think this is ok, we can just add a warning that seeds should be expected to be used within a version.

Reply:
Copying over a message I sent to @ekolve for completeness. Feel free to chime in.
TL;DR: I don't think random seed should be set inside of any action parameters. It should only be set globally.

----------------

So I have a problem with randomSeed at the moment. Suppose I have a loop:

```python
controller = Controller()
for episode in range(int(1e8)):
    controller.step(action="RandomizeMaterials", randomSeed=42)
    # Do something...
```

this has me expecting the materials will change once, and then never change after that because the seed is reset each time? (Here, from C#, you'd set Random.Seed = 42).That is how randomSeed is used in other actions, but this is very unnatural and basically introducing a silent bug, as one would expect calling RandomizeMaterials many times inside of this loop would constantly change the materials differently.I think there should instead be something like:

```python
controller = Controller(randomSeed=42)
for episode in range(int(1e8)):
    controller.step(action="RandomizeMaterials")
```

or

```python
controller = Controller()
controller.randomSeed = 42
for episode in range(int(1e8)):
    controller.step(action="RandomizeMaterials")
```

which is done in other libraries, which sets the random seed for everything up front, and then samples from that random seed later. But that might be beyond the scope of this PR.

Reply:
Setting a global random seed would lead to a lot of confusion about to reproduce a configuration.  If the seed is set once on startup, then all the actions that depend on the RNG must be called in the *exact* same order otherwise the configuration will be different.  Whats complicates things more is suppose that you don't want to call something like 'PutObject' (which has randomization) for a particular scene setup, then you basically won't be able to have the same seed for a subsequent call to another action that the RNG depends on.  As well, if the seed is set once for the Player, then all training, scene loading, etc. must occur in exactly the same order and each action must make the exact same number of calls to the RNG or a subsequent's action won't receive the same sequence of random numbers.

Reply:
After further discussion, we are moving this to a future PR.